src/coreComponents/constitutive/fluid/chemicalReactions/EquilibriumReactions.cpp

```diff
@@ -0,0 +1,339 @@
+  // Second index: 0 = H+, 1 = HCO3-, 2 = Ca+2, 3 = SO4-2, 4 = Cl-, 5 = Mg+2, 6 = Na+1
+  m_stoichMatrix.resize( m_numSecondarySpecies, m_numPrimarySpecies );
+  m_stoichMatrix[0][0] = -1;
+  m_stoichMatrix[1][0] = 1;
+  m_stoichMatrix[1][1] = 1;
+  m_stoichMatrix[2][0] = -1;
+  m_stoichMatrix[2][1] = 1;
+  m_stoichMatrix[3][0] = 1;
+  m_stoichMatrix[3][1] = 1;
+  m_stoichMatrix[4][1] = 1;
+  m_stoichMatrix[4][2] = 1;
+  m_stoichMatrix[5][0] = -1;
+  m_stoichMatrix[5][1] = 1;
+  m_stoichMatrix[5][2] = 1;
+  m_stoichMatrix[6][2] = 1;
+  m_stoichMatrix[6][3] = 1;
+  m_stoichMatrix[7][2] = 1;
+  m_stoichMatrix[7][4] = 1;
+  m_stoichMatrix[8][2] = 1;
+  m_stoichMatrix[8][4] = 2;
+  m_stoichMatrix[9][5] = 1;
+  m_stoichMatrix[9][3] = 1;
+  m_stoichMatrix[10][6] = 1;
+  m_stoichMatrix[10][3] = 1;
+
+  // Equilibrium Constant
+  m_log10EqConst.resize( m_numSecondarySpecies );
+  m_log10EqConst[0] = 13.99;
+  m_log10EqConst[1] = -6.36;
+  m_log10EqConst[2] = 10.33;
+  m_log10EqConst[3] = -3.77;
+  m_log10EqConst[4] = -1.09;
+  m_log10EqConst[5] = 7.07;
+  m_log10EqConst[6] = -2.16;
+  m_log10EqConst[7] = 0.67;
+  m_log10EqConst[8] = 0.60;
+  m_log10EqConst[9] = -2.43;
+  m_log10EqConst[10] = -0.82;
+}
+
+EquilibriumReactions::KernelWrapper EquilibriumReactions::createKernelWrapper() const
+{
+  return KernelWrapper( m_numPrimarySpecies,
+                        m_numSecondarySpecies,
+                        m_log10EqConst,
+                        m_stoichMatrix,
+                        m_chargePrimary,
+                        m_chargeSec,
+                        m_ionSizePrimary,
+                        m_ionSizeSec,
+                        m_DebyeHuckelA,
+                        m_DebyeHuckelB,
+                        m_WATEQBDot );
+}
+
+
+GEOSX_HOST_DEVICE
+void EquilibriumReactions::KernelWrapper::updateConcentrations( real64 const temperature,
+                                                                arraySlice1d< real64 const, compflow::USD_COMP - 1 > const & primarySpeciesTotalConcentration,
+                                                                arraySlice1d< real64, compflow::USD_COMP - 1 > const & primarySpeciesConcentration,
+                                                                arraySlice1d< real64, compflow::USD_COMP - 1 > const & secondarySpeciesConcentration ) const
+
+{
+  array2d<real64> matrix( m_numPrimarySpecies, m_numPrimarySpecies );
+  array1d<real64> rhs( m_numPrimarySpecies );
+  array1d<real64> solution( m_numPrimarySpecies );
+
+  setInitialGuess(primarySpeciesTotalConcentration, primarySpeciesConcentration);
+  std::cout<<"Initial guess " << primarySpeciesConcentration <<std::endl;
+  std::cout<<"Total concentration to solve for " << primarySpeciesTotalConcentration <<std::endl;
+
+
+  bool converged = false;
+  for( int iteration = 0; iteration < m_maxNumIterations; iteration++ )
```

Comment:
There is also a Newton solver [here](https://github.com/GEOSX/GEOSX/blob/develop/src/coreComponents/constitutive/fluid/PVTFunctions/CO2EOSSolver.cpp). It seems a little bit difficult to use it in the new `EquilibriumReactions` class, but maybe later we can unify them, or at least group them in the same place.

Reply:
ah, yeah, we should really write a generic nonllinear solver.